### ISSUE TITLE
[HOLD for payment 2022-04-13] [HOLD for payment 2022-04-12] Audit forms and fix inconsistencies with focus, tab and shift + tab behavior 1/4

### ISSUE THREAD

**Issue by luacmartins** [Expensify/App#7523]
We should audit all our forms and fix any inconsistencies with focus, tab, shift + tab and enter behavior. The expected behavior is as follows:

1. `Tab` navigates to the next input.
2. `Shift + tab` navigates to the previous input.
3. `Enter` submits the form.

**Note:** We should make sure that tabbing cycles through the form in an order that makes sense, usually top to bottom.

Here's a list of forms to be audited:

- [ ] [AddPlaidBankAccount.js](https://github.com/Expensify/App/blob/main/src/components/AddPlaidBankAccount.js)
- [ ] [ACHContractStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/ACHContractStep.js)
- [ ] [BankAccountStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/BankAccountStep.js)
- [ ] [CompanyStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/CompanyStep.js)
- [ ] [Password form](https://github.com/Expensify/App/blob/main/src/pages/signin/PasswordForm.js)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @trjExpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by luacmartins** [comment]
Actually, we might use Applause to do this audit for us. Removing the External label.

**Comment by MelvinBot** [comment]
12 days overdue. Walking. Toward. The. Light...

**Comment by MelvinBot** [comment]
12 days overdue now... This issue's end is nigh!

**Comment by luacmartins** [comment]
Nvm my previous comment, this can be handled by contributors! Adding `External` label.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @dylanexpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by luacmartins** [comment]
@dylanexpensify please note that we should add one milestone per 5 forms (4 milestones total).

**Comment by dylanexpensify** [comment]
Sounds great! Will get on this now @luacmartins !

**Comment by dylanexpensify** [comment]
discussing in slack best way to approach posting this issue on upwork!

**Comment by dylanexpensify** [comment]
not overdue - going to post tomorrow by breaking this up into smaller bites instead of milestones

**Comment by dylanexpensify** [comment]
posting breakout today!

**Comment by dylanexpensify** [comment]
Broke down the issue, posting this to Upwork now!

**Comment by dylanexpensify** [comment]
Internal: https://www.upwork.com/ab/applicants/1497211614476763136/job-details
External: https://www.upwork.com/jobs/~01d7fed60e5d35c64d

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @thienlnam (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by Santhosh-Sellavel** [comment]
## Proposal: 
I am interested in this!

I will perform the audit for each page for the actions stated in the post [here](https://github.com/Expensify/App/issues/7523#issue-1122232308)
Find out the inconsistency if any and will fix them.

**Comment by Santhosh-Sellavel** [comment]
Few forms might not be accessible to contributors. Is there any way to open those forms directly?

**Comment by rushatgabhane** [comment]
@Santhosh-Sellavel  if you're having cors issue on web, use this extension - https://chrome.google.com/webstore/detail/cors-unblock/lfhmikememgdcahcdlaciloancbhjino?hl=en

And here are the credentials for Plaid
> Add a bank account by going to Settings > Payments >  Add Payment Method > Bank Account

> Select Wells Fargo Bank Account
 Use credentials: user_good, pass_good

**Comment by Santhosh-Sellavel** [comment]
@luacmartins 

These forms, I'm not sure how can contributor open these forms,

1. Is this Form accessed via `Settings > Payments > Add Payment Method > Bank Account` ?
 [AddPlaidBankAccount.js](https://github.com/Expensify/App/blob/main/src/components/AddPlaidBankAccount.js)
 
2. If I am right these forms are part of `Connect bank account flow in Workspace.`
 [BankAccountStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/BankAccountStep.js)
 [CompanyStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/CompanyStep.js)
 
3. Where do these steps occur?
 [ACHContractStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/ACHContractStep.js)
 
 We need to add a valid bank account to proceed with forms mention 1 & 2 points.

**Comment by Santhosh-Sellavel** [comment]
> Add a bank account by going to Settings > Payments > Add Payment Method > Bank Account
> 
> Select Wells Fargo Bank Account
> Use credentials: user_good, pass_good

@rushatgabhane This does not work for me I tried on staging web app.

**Comment by luacmartins** [comment]
@Santhosh-Sellavel here's how to access these forms:

1. [AddPlaidBankAccount.js](https://github.com/Expensify/App/blob/main/src/components/AddPlaidBankAccount.js) - `Settings > Workspace > Connect bank account > Connect online with Plaid`
2. [BankAccountStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/BankAccountStep.js), [CompanyStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/CompanyStep.js) and [ACHContractStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/ACHContractStep.js) are all part of the `Connect bank account flow in Workspace`, accessed by the flow mentioned in 1.

>This does not work for me I tried on staging web app.

These credentials are valid for `dev` only. They will not work on `staging`.

**Comment by rushatgabhane** [comment]
Thanks for the help @ luacmartins

**Comment by rushatgabhane** [comment]
🎀👀🎀 C+ reviewed

@Santhosh-Sellavel wants to volunteer for the audit, let's get this party started!
cc: @thienlnam

**Comment by thienlnam** [comment]
@Santhosh-Sellavel Do you feel as if you are able to access the forms now?  Since you have volunteered for it you can have the job 🟢

**Comment by MelvinBot** [comment]
📣 @Santhosh-Sellavel You have been assigned to this job by @thienlnam!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by Santhosh-Sellavel** [comment]
Still have some issues with adding a bank account.

Requested help on slack [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1646164482111669?thread_ts=1646163503.484199&cid=C01GTK53T8Q)

cc: @thienlnam @rushatgabhane

**Comment by luacmartins** [comment]
@Santhosh-Sellavel glad that I could help getting access to the forms. It seems that [AddPlaidBankAccount.js](https://github.com/Expensify/App/blob/main/src/components/AddPlaidBankAccount.js) will have to be audited internally though.

**Comment by Santhosh-Sellavel** [comment]
@luacmartins  or @rushatgabhane 
Have a look at the below video. I am performing `Tab` and `Shift + Tab` operations. After reaching the First or Input at the top. When I perform `Shift + Tab`, the focus changes to the close button. This is expected behavior right?

https://user-images.githubusercontent.com/85645967/156248891-db7c2076-f56a-4dda-99be-0754ace2e2dc.mov

**Comment by rushatgabhane** [comment]
Yes, looks good @Santhosh-Sellavel

**Comment by Santhosh-Sellavel** [comment]
In other words, the goal of the task is to make sure that performing `tab` &` shift + tab` operations ensure that the focus is switching to the next or previous input in focus cycles the top to bottom or bottom to top in the right orders. Once the end of When First or Last Input is reached by performing `tab` operations, the focus could switch to the next view that may require focus.

**Comment by rushatgabhane** [comment]
Yep, and the order is from Top to Bottom and Left to Right for `Tab`

**Comment by Santhosh-Sellavel** [comment]
@rushatgabhane @luacmartins I performed the audit forms couple of times. For all the forms, tab cycles work in the right order.

**Comment by thienlnam** [comment]
Since there isn't a PR associated with this, instead could you just record a video going through each form and demonstrating that these work as expected?  That way, we can consider it completed

**Comment by Santhosh-Sellavel** [comment]
I'll update the videos within a couple of days.  
Also, I am sure this will require a PR.

**Comment by mdneyazahmad** [comment]
@Santhosh-Sellavel Did you confirm the tab cycles through all input and does not focus element outside the open modal.

Let's say we have 3 input. When you `tab` and reach the last and now when you press `tab` you should focus on the first input and not on the element outside of the modal. Same is the case when `shift + tab`.

**Comment by mdneyazahmad** [comment]
## Proposal
These 4 related issues are very similar. I proposed the solution here https://github.com/Expensify/App/issues/7918#issuecomment-1063298280

**Comment by rushatgabhane** [comment]
> When you tab and reach the last and now when you press tab you should focus on the first input and not on the element outside of the modal.

Wait, really? I don't think we should do that. This will break accessibility features when using tab.

**Comment by Santhosh-Sellavel** [comment]
cc: @luacmartins Can you help us here.

**Comment by mdneyazahmad** [comment]
@rushatgabhane This is the only case when we open a modal. When we close it everything goes to the normal state. Here is the reference https://www.w3.org/TR/wai-aria-practices/#dialog_modal and demo https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html

**Comment by rushatgabhane** [comment]
@mdneyazahmad I believe we don't have a modal for the pages listed in this issue.

**Comment by mdneyazahmad** [comment]
@rushatgabhane You may be right and I misinterpreted the requirements.

**Comment by rushatgabhane** [comment]
No worries, thanks for your input on this! :)

**Comment by luacmartins** [comment]
I agree with @parasharrajat! The RHN blocks usage of the remainder of the app until it's dismissed. We could benefit from focus trapping here.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @michaelhaxhiu (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by dylanexpensify** [comment]
Reassigning as I'm OOO next week! Thanks Hax! I'll pick it up when I'm back!

**Comment by thienlnam** [comment]
The focus trapping will affect all of these forms though instead of holding on that, I think we should have contributors audit these forms in parallel and then make another issue that goes through each of the forms and adds focus trapping.  Thoughts @luacmartins?

**Comment by luacmartins** [comment]
I agree! Let's stick to auditing the forms for now. We might be changing the way the RHN is displayed soon anyways.

**Comment by thienlnam** [comment]
@Santhosh-Sellavel Please proceed with the form audits

**Comment by Santhosh-Sellavel** [comment]
- [ ] [AddPlaidBankAccount.js](https://github.com/Expensify/App/blob/main/src/components/AddPlaidBankAccount.js)
- [ ] [ACHContractStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/ACHContractStep.js)
- [ ] [BankAccountStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/BankAccountStep.js)
- [ ] [CompanyStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/CompanyStep.js)
- [ ] [Password form](https://github.com/Expensify/App/blob/main/src/pages/signin/PasswordForm.js)

**Comment by Santhosh-Sellavel** [comment]
[AddPlaidBankAccount.js](https://github.com/Expensify/App/blob/main/src/components/AddPlaidBankAccount.js)

https://user-images.githubusercontent.com/85645967/157957633-e5282109-2ed7-477d-8876-21a3b0caee10.mov

1. Focus on to eye symbol is the issue here right? @luacmartins 
Everything else LGMT

**Comment by Santhosh-Sellavel** [comment]
[BankAccountStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/BankAccountStep.js)
No issues, LGTM

https://user-images.githubusercontent.com/85645967/157958499-48fa822f-9c92-4618-b91c-3aa9a2ac01ba.mov

**Comment by Santhosh-Sellavel** [comment]
[Password form](https://github.com/Expensify/App/blob/main/src/pages/signin/PasswordForm.js)
Except - Show Password eye symbol, everything else LGTM

https://user-images.githubusercontent.com/85645967/157958975-33112962-d3b3-4980-b35f-d73c673114ea.mov

**Comment by Santhosh-Sellavel** [comment]
[CompanyStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/CompanyStep.js)

Everything LGTM.

Except, not sure on this Date Field, focus changes to `dd` to `mm` to `yyyy`

## TAB till the end of the form
https://user-images.githubusercontent.com/85645967/157959549-06760eba-abce-4ff1-bfd6-9e19f1dead05.mov

## SHIFT + TAB till top of the form & Enter
https://user-images.githubusercontent.com/85645967/157960089-f2de9d74-9b6c-4f42-9061-4de6e0238095.mov

**Comment by Santhosh-Sellavel** [comment]
[ACHContractStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/ACHContractStep.js)

Issue
1. With checkbox focused pressing enter submits the form & also inverts the checkbox selection.
Everything else LGTM.

### TAB & SHIFT + TAB Cycle Without `Somebody else owns more than 25% of` Form
https://user-images.githubusercontent.com/85645967/157962722-ef0747d4-c205-480c-b8c6-679298931adf.mov

### TAB & SHIFT + TAB Cycle for `Somebody else owns more than 25% of` Form
https://user-images.githubusercontent.com/85645967/157963919-5e9a54b3-3103-43c4-a7c8-f95d1f5d6104.mov

**Comment by Santhosh-Sellavel** [comment]
Hi @thienlnam @luacmartins,

I posted a separate comment for each form and posted my thoughts specific to that form like what I felt was an issue and specific questions too. In each form's screen recording, I perform `TAB` until focus reaches the END of the form & `SHIFT + TAB` until the focus reaches the top of the form finally `ENTER` to submit the form.

I have a couple of questions to clarify since this is common to all forms.
1. Should we tab cycles navigate focus only through input/checkboxes or can it focus on other elements too like buttons, links, text, etc?
2. After opening a form, initially focus will not be set to any inputs. So if the user press `TAB` should we focus on the first input of the form?

All forms & corresponding comments are linked below for easier navigation.
[AddPlaidBankAccount.js](https://github.com/Expensify/App/blob/main/src/components/AddPlaidBankAccount.js) -> https://github.com/Expensify/App/issues/7523#issuecomment-1065495907
[ACHContractStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/ACHContractStep.js) -> https://github.com/Expensify/App/issues/7523#issuecomment-1065520027
[BankAccountStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/BankAccountStep.js) -> https://github.com/Expensify/App/issues/7523#issuecomment-1065497572
[CompanyStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/CompanyStep.js) -> https://github.com/Expensify/App/issues/7523#issuecomment-1065512322
[Password form](https://github.com/Expensify/App/blob/main/src/pages/signin/PasswordForm.js) -> https://github.com/Expensify/App/issues/7523#issuecomment-1065500916

Please let me know if there are any other issues, I missed out on. Thanks!

cc: @rushatgabhane

**Comment by luacmartins** [comment]
https://github.com/Expensify/App/issues/7523#issuecomment-1065495907 - I think we should keep the focus on the show password button for accessibility. Could we make it more obvious though? I feel like the left border seems off.

https://github.com/Expensify/App/issues/7523#issuecomment-1065520027 - Hmm I think we should toggle the checkbox on "Enter" when it has focus, instead of submitting the form. Additionally, I think the checkbox and text should be focused at the same time.

https://github.com/Expensify/App/issues/7523#issuecomment-1065497572 - Same comment above about the text and checkbox being focused separately (this does not apply to the link though, as that one should be focused separately in case a user wants to click it). What happens when we press "Enter" with the checkbox in focus?

https://github.com/Expensify/App/issues/7523#issuecomment-1065512322 - Same comment as above about the checkbox. As for date picker, I think that we should remove tab from within it, i.e. pressing tab should move us to the next input.

https://github.com/Expensify/App/issues/7523#issuecomment-1065500916 - I think we should keep the focus on the show password button.

**Comment by Santhosh-Sellavel** [comment]
> What happens when we press "Enter" with the checkbox in focus?

The checkbox is getting selected & immediately form also gets submitted!

> https://github.com/Expensify/App/issues/7523#issuecomment-1065520027 - Hmm I think we should toggle the checkbox on "Enter" when it has focus, instead of submitting the form. Additionally, I think the checkbox and text should be focused at the same time.

I believe using the "Space" key would make the most obvious choice to toggle between selections in the checkbox.  That’s the commonly used way.

Play with checkbox here -> https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox

**Comment by luacmartins** [comment]
>The checkbox is getting selected & immediately form also gets submitted!

Hmm if we are using "Space" to toggle the checkbox, then pressing "Enter" should submit the form and I'd think that this is expected?

>I believe using the "Space" key would make the most obvious choice to toggle between selections in the checkbox. That’s the commonly used way.

Yes, that makes sense. We should add that if it's not currently a feature.

**Comment by Santhosh-Sellavel** [comment]
Not yet started to work on the fixes, will provide an update on this next week!

**Comment by Santhosh-Sellavel** [comment]
@luacmartins 
 
#### Form 1 && Form 5
 
> https://github.com/Expensify/App/issues/7523#issuecomment-1065495907 - I think we should keep the focus on the show password button for accessibility. Could we make it more obvious though? I feel like the left border seems off.
> https://github.com/Expensify/App/issues/7523#issuecomment-1065500916 - I think we should keep the focus on the show password button.

#### Solution
The height of the `showPassword` button and password input is the same. Since it is rendered inside the container except left all other borders got cut off

Applying 1px margin and applying border-radius same as the container would look like this.

<img width="367" alt="Screenshot 2022-03-23 at 11 53 43 PM" src="https://user-images.githubusercontent.com/85645967/159769818-711fdc4e-bd21-4ca8-877b-b2f73b28a545.png">
___________________________________________________

#### Form 2 && Form 3
> https://github.com/Expensify/App/issues/7523#issuecomment-1065520027 - Hmm I think we should toggle the checkbox on "Enter" when it has focus, instead of submitting the form. Additionally, I think the checkbox and text should be focused at the same time.
> https://github.com/Expensify/App/issues/7523#issuecomment-1065497572 - Same comment above about the text and checkbox being focused separately (this does not apply to the link though, as that one should be focused separately in case a user wants to click it). What happens when we press "Enter" with the checkbox in focus?

#### Solution:
I proposed to use space for check selection. Discussed on the 
https://github.com/Expensify/App/issues/7523#issuecomment-1068209007
https://github.com/Expensify/App/issues/7523#issuecomment-1068548722

To use `Space bar` for selection is separate a feature request, I propose we can handle this as a separate feature request, not with this issue. If you agree with me I will post it on slack or let's create a ticket on GH itself. 

The above will address the Enter conflict behavior.

________________________________________________________________

### Form 4
> https://github.com/Expensify/App/issues/7523#issuecomment-1065512322 - Same comment as above about the checkbox. As for the date picker, I think that we should remove the tab from within it, i.e. pressing the tab should move us to the next input.

For Date Picker, I believe working as expected. By default, it's allowed to navigate through all fields for a date the reference here => https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date.
____________________________________________________________________

#### Discussion
About,
> I think the checkbox and text should be focused at the same time.

Checkbox & Text has separate onPress. So focusing on them separately makes sense. If we want to achieve the focus to both we can do this by wrapping `Text` & `Checkbox` inside a `Pressable` instead of `Text` alone [here](https://github.com/Expensify/App/blob/116398b0d2195e43c185ee72e70b214def73e410/src/components/CheckboxWithLabel.js#L74-L104).  But still it focuses the Checkbox alone separately after focusing the both, Checkbox alone is still clickable.


Please post your thoughts on this!
cc: @thienlnam @rushatgabhane

**Comment by luacmartins** [comment]
Thanks for the proposed solutions @Santhosh-Sellavel! Here are my thoughts:

**Form 1 && Form 5**
Can we remove the gap between borders somehow? It looks a bit off.

**Form 2 && Form 3**
Sounds good! Let's do it.

**Form 4**
I agree. I think it would make sense to remove the tabbing if we didn't have the input pattern defined there. Let's do nothing here.

>Checkbox & Text has separate onPress. So focusing on them separately makes sense. If we want to achieve the focus to both we can do this by wrapping Text & Checkbox inside a Pressable instead of Text alone [here](https://github.com/Expensify/App/blob/116398b0d2195e43c185ee72e70b214def73e410/src/components/CheckboxWithLabel.js#L74-L104). But still it focuses the Checkbox alone separately after focusing the both, Checkbox alone is still clickable.

On second thought, let's keep it as is.

**Comment by Santhosh-Sellavel** [comment]
> Form 1 && Form 5
Can we remove the gap between borders somehow? It looks a bit off.

How does this look?
<img width="213" alt="Screenshot 2022-03-28 at 11 07 38 PM" src="https://user-images.githubusercontent.com/85645967/160455452-ed65f0a2-8630-40ed-8015-1142d1d45349.png">

**Comment by luacmartins** [comment]
Looks good! Thanks!

**Comment by Santhosh-Sellavel** [comment]
PR is up!

**Comment by rushatgabhane** [comment]
Hmm, I've verified the audit. But I didn't help here in any other way 😅
Sorry about that. 

The PR was straightforward. I'm gonna unassign myself as a C+

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.49-1 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/8346


If no regressions arise, payment will be issued on 2022-04-12. :confetti_ball:

**Comment by dylanexpensify** [comment]
@Santhosh-Sellavel we don't have any applicants for the Upwork job cc @michaelhaxhiu. Who should be applying/paid?

**Comment by dylanexpensify** [comment]
Great! Getting job reposted for @Santhosh-Sellavel to apply, will then pay out!

**Comment by dylanexpensify** [comment]
Repost

Internal: https://www.upwork.com/ab/applicants/1514144553994493952/job-details
External: https://www.upwork.com/jobs/~011512c770333f888a

cc @Santhosh-Sellavel

**Comment by dylanexpensify** [comment]
@Santhosh-Sellavel offer sent!

**Comment by dylanexpensify** [comment]
Payment sent, contract ended, post removed!
